### PR TITLE
API-breaking changes planned for 0.7

### PR DESCRIPTION
**Pull request by abergeron**

### CODE REVIEW

File: src/gpuarray/buffer.h [hunk 1]
```diff
@@ -550,13 +550,7 @@ GPUARRAY_PUBLIC gpucontext *gpukernel_context(gpukernel *k);
  */
 #define GA_CTX_PROP_DEVNAME  1
 
-/**
```
Comment (nouiz) [inline src/gpuarray/buffer.h]: why remove them? I think they are still needed.

Comment (abergeron) [inline src/gpuarray/buffer.h]: This is exactly the same as GA_CTX_PROP_MAXLSIZE0.  So I want to remove this alias since it may be confused for the maximum total local size.

File: src/cluda_opencl.h.c [hunk 2]
```diff
@@ -0,0 +1,107 @@
+static const char cluda_opencl_h[] = {
```
Comment (nouiz) [inline src/cluda_opencl.h.c]: what is this?

File: src/cluda_opencl.h [hunk 3]
```diff
@@ -0,0 +1,40 @@
+#define local_barrier() barrier(CLK_LOCAL_MEM_FENCE)
```
Comment (nouiz) [inline src/cluda_opencl.h]: I think it would be better to wrap this into something like:
```
#ifndef CLUDA_H
...
#define CLUDA_H
#endif
```

File: src/head.py [hunk 4]
```diff
@@ -0,0 +1,35 @@
+def wrt(f, n, b):
```
Comment (nouiz) [inline src/head.py]: Add comment to tell what this file does and when it should be used.

File: src/gpuarray_buffer_cuda.c [hunk 5]
```diff
@@ -322,7 +322,8 @@ cuda_context *cuda_make_ctx(CUcontext ctx, gpucontext_props *p) {
   }
   res->errbuf->flags |= CUDA_MAPPED_PTR;
   /* Prime the cache */
-  cuda_alloc((gpucontext *)res, p->initial_cache_size, NULL, 0);
+  if (p->initial_cache_size)
```
Comment (nouiz) [inline src/gpuarray_buffer_cuda.c]: I forgot the detail, but why this don't cause a memory leak? We want to just fill the cache, not use it.

File: tests/check_array.c [hunk 6]
```diff
@@ -27,20 +27,20 @@ START_TEST(test_take1_ok) {
                               18, 19, 20, 21, 22, 23};
   uint32_t buf[12 * 24];
   const size_t data_dims[1] = {24};
-  ssize_t indexes[12];
```
Comment (nouiz) [inline tests/check_array.c]: Why did you remove the ability to use those dtype on the GPU?

File: src/CMakeLists.txt [hunk 7]
```diff
@@ -14,6 +14,22 @@ add_custom_command(
   WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}
   DEPENDS ${CMAKE_CURRENT_SOURCE_DIR}/gen_types.py)
 
+add_custom_command(
```
Comment (nouiz) [inline src/CMakeLists.txt]: Why do you continue to keep them in the repo if you regenete them automatically? I would prefer to remove them from the repo.

File: pygpu/gpuarray.pxd [hunk 8]
```diff
@@ -172,10 +172,6 @@ cdef extern from "gpuarray/array.h":
                           gpudata *data, size_t offset, int typecode,
                           unsigned int nd, const size_t *dims,
                           const ssize_t *strides, int writable)
-    int GpuArray_copy_from_host(_GpuArray *a,
```
Comment (nouiz) [inline pygpu/gpuarray.pxd]: What was the bug? crash? bad values?

File: pygpu/reduction.py [hunk 9]
```diff
@@ -37,7 +37,8 @@ def _ceil_log2(x):
 
 #define REDUCE(a, b) (${reduce_expr})
 
-KERNEL void ${name}(const unsigned int n, ${out_arg.decltype()} out
+KERNEL void ${name}(const unsigned int n, ${out_arg.decltype()} out,
```
Comment (nouiz) [inline pygpu/reduction.py]: why the offset is needed on the output?

File: src/gpuarray_buffer_cuda.c [hunk 10]
```diff
@@ -37,6 +37,9 @@ STATIC_ASSERT(sizeof(GpuArrayIpcMemHandle) == sizeof(CUipcMemHandle), cuda_ipcme
  */
 #define FRAG_SIZE (64)
 
+extern gpuarray_blas_ops cublas_ops;
```
Comment (abergeron) [inline src/gpuarray_buffer_cuda.c]: I don't want to expose it as a property anymore.

File: src/gpuarray_buffer_cuda.c (hunk 5 above)
Comment (abergeron) [inline src/gpuarray_buffer_cuda.c]: Yes, I'll fix it.

File: src/gpuarray_buffer_cuda.c [hunk 11]
```diff
@@ -758,7 +776,7 @@ static gpudata *cuda_alloc(gpucontext *c, size_t size, void *data, int flags) {
    * to a multiple of FRAG_SIZE.  This also ensures that if we split a
    * block, the next block starts properly aligned for any data type.
    */
-  if (!(ctx->flags & GA_CTX_DISABLE_ALLOCATION_CACHE)) {
+  if (ctx->max_cache_size != 0) {
```
Comment (abergeron) [inline src/gpuarray_buffer_cuda.c]: max_cache_size = SIZE_T_MAX (or anything bigger than the GPU memory), initial_cache_size = 0

File: src/gpuarray_buffer.c [hunk 12]
```diff
@@ -37,15 +37,84 @@ int gpu_get_device_count(const char* name, unsigned int platform,
...
+  return GA_NO_ERROR;
+}
+
+int gpucontext_props_sched(gpucontext_props *p, int sched) {
+  if (sched == GA_CTX_SCHED_MULTI)
+    FLSET(p->flags, GA_CTX_MULTI_THREAD);
+  else
+    FLCLR(p->flags, GA_CTX_MULTI_THREAD);
+
+  switch (sched) {
```
Comment (abergeron) [inline src/gpuarray_buffer.c]: ok

File: src/gpuarray_buffer.c [hunk 13]
```diff
@@ -37,15 +37,84 @@ int gpu_get_device_count(const char* name, unsigned int platform,
...
-  return res;
+  gpucontext *r;
+  if (ops == NULL) {
+    gpucontext_props_del(p);
+    return global_err->code;
+  }
+  if (p == NULL && gpucontext_props_new(&p) != GA_NO_ERROR)
+    return global_err->code;
+  r = ops->buffer_init(p);
+  gpucontext_props_del(p);
```
Comment (abergeron) [inline src/gpuarray_buffer.c]: This is an ease of use concern.  I wouldn't make that the default way objects are handled, but I think this is a reasonable exception.

You could say that once you pass the props object to gpucontext_init() it becomes the property of the context and it will manage its (very short) lifetime.

File: src/cluda_opencl.h.c (hunk 2 above)
Comment (abergeron) [inline src/cluda_opencl.h.c]: This is to embed the source code in cluda_opencl.h in the binary so that we can use it for kernels.

File: src/cluda_opencl.h (hunk 3 above)
Comment (abergeron) [inline src/cluda_opencl.h]: ok

File: tests/check_array.c (hunk 6 above)
Comment (abergeron) [inline tests/check_array.c]: That is groundwork for call32.  Also I didn't remove all uses of the type, just the ability to make arrays, because arrays of a type that can change size is a bad idea.

File: src/CMakeLists.txt (hunk 7 above)
Comment (abergeron) [inline src/CMakeLists.txt]: This is intended for developers that will modify the .h files directly.  If you just build from the source, it shouldn't have to regenerate since the files are unmodified.

File: pygpu/gpuarray.pxd (hunk 8 above)
Comment (abergeron) [inline pygpu/gpuarray.pxd]: There was some bad values, but it's also an ill-concieved idea from the start.

The idea was to avoid copying data on the CPU during a transfer so it would figure out the start and the end of the passed-in data (using strides and shapes) and copy the whole block.  This would potentially copy large swaths of irrelevant data that just happen to sit between those boundaries.

It also bypassed the forced offset that is added in DEBUG and create some problems related to that.

File: pygpu/reduction.py (hunk 9 above)
Comment (abergeron) [inline pygpu/reduction.py]: Because it's an array and therefore it can have an offset.

File: src/gpuarray/buffer.h (hunk 1 above)
Comment (nouiz) [inline src/gpuarray/buffer.h]: I got confused myself! So what is the name for the maximum total local size? Is it available?

File: src/gpuarray_buffer.c (hunk 13 above)
Comment (nouiz) [inline src/gpuarray_buffer.c]: ok, but document that well as this is non-standard.

This type of thing can be good, but can also cause problems as some non-stardard stuff in the NumPy C API.

File: tests/check_array.c (hunk 6 above)
Comment (nouiz) [inline tests/check_array.c]: ok

File: src/head.py (hunk 4 above)
Comment (nouiz) [inline src/head.py]: Can you add how to use that script in the comment? Put the command line you use yourself.

**Comment by nouiz** [comment]
All CI failed

File: src/gpuarray/buffer.h (hunk 1 above)
Comment (abergeron) [inline src/gpuarray/buffer.h]: There isn't one currently.  But we can expose it if required.

File: src/gpuarray_buffer_cuda.c (hunk 11 above)
Comment (abergeron) [inline src/gpuarray_buffer_cuda.c]: The theano PR does this correctly now, since the problem was in pygpu.

I believe there is documentation about what max_cache_size and initial_cache_size mean, and if not, I should add it.  But I wouldn't explicitly document this combination since it follows naturally from the properties of both limits.

File: src/gpuarray_buffer.c (hunk 13 above)
Comment (abergeron) [inline src/gpuarray_buffer.c]: I'll add some documentation about that.

File: src/head.py (hunk 4 above)
Comment (abergeron) [inline src/head.py]: ok

File: src/gpuarray_buffer_cuda.c (hunk 11 above)
Comment (abergeron) [inline src/gpuarray_buffer_cuda.c]: Ok there is no documentation, but I'll add some.

**Comment by nouiz** [comment]
I'm not able to run libgpuarray tests:
```
.../repos/libgpuarray/Build$ DEVICE=cuda0 make test
Running tests...
Test project /u/bastienf/repos/libgpuarray/Build
No tests were found!!!
```

this was working before. This is taken from doc/installation.txt

File: src/gpuarray/buffer.h (hunk 1 above)
Comment (nouiz) [inline src/gpuarray/buffer.h]: ok

File: src/gpuarray_buffer_cuda.c [hunk 14]
```diff
@@ -322,8 +322,11 @@ cuda_context *cuda_make_ctx(CUcontext ctx, gpucontext_props *p) {
   }
   res->errbuf->flags |= CUDA_MAPPED_PTR;
   /* Prime the cache */
-  if (p->initial_cache_size)
-    cuda_alloc((gpucontext *)res, p->initial_cache_size, NULL, 0);
+  if (p->initial_cache_size) {
+    gpudata *tmp = cuda_alloc((gpucontext *)res, p->initial_cache_size, NULL, 0);
```
Comment (nouiz) [inline src/gpuarray_buffer_cuda.c]: IRL answer, opencl don't have a cache

**Comment by nouiz** [comment]
There is still a usage of a function you removed:

tests/check_collectives.c:  err = GpuArray_copy_from_host(&Adev, ctx, A, GA_INT, ND, indims, instrds); \

**Comment by nouiz** [comment]
@abergeron told that jenkins probably don't have MPI installed, so didn't compiled collective tests.

@slefrancois when you have time, add that to jenkins installation.
